# OLAP Cubes

## 1. Slicing and Dicing

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by creating and connecting to the database

In [1]:
!pwd

/home/johnlin/python_repos/data_engineer_related/0.basics/4_Data Warehouses


### Connect to the local database where Pagila is loaded

In [1]:

%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'student'
DB_PASSWORD = 'example'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:example@127.0.0.1:5432/pagila


### Star Schema

<img src="pagila-star.png" width="50%"/>

### Starting with a simple cube
Calculate the revenue (sales_amount) by day, rating, and city. Then, sort by revenue in descending order and limit to the first 20 rows.

In [2]:
%%time
%%sql

SELECT d.day AS day, m.rating AS rating, c.city AS city, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
GROUP BY day, rating, city
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila
20 rows affected.
CPU times: user 8.09 ms, sys: 0 ns, total: 8.09 ms
Wall time: 86.8 ms


[(15, 'G', 'Kermanshah', Decimal('23.96')),
 (14, 'PG-13', 'Brockton', Decimal('21.97')),
 (15, 'PG-13', 'Jhansi', Decimal('20.97')),
 (4, 'NC-17', 'Lapu-Lapu', Decimal('19.97')),
 (18, 'R', 'Qomsheh', Decimal('19.97')),
 (2, 'NC-17', 'Shubra al-Khayma', Decimal('18.97')),
 (4, 'R', 'Siegen', Decimal('18.96')),
 (1, 'PG', 'Izumisano', Decimal('18.96')),
 (17, 'R', 'Athenai', Decimal('17.98')),
 (24, 'PG', 'Talavera', Decimal('17.98')),
 (3, 'PG', 'Cape Coral', Decimal('17.98')),
 (24, 'PG', 'Memphis', Decimal('17.98')),
 (13, 'G', 'Cape Coral', Decimal('17.97')),
 (10, 'PG-13', 'Uruapan', Decimal('17.97')),
 (8, 'PG-13', 'Probolinggo', Decimal('16.98')),
 (23, 'G', 'Mukateve', Decimal('16.98')),
 (1, 'PG-13', 's-Hertogenbosch', Decimal('16.98')),
 (14, 'NC-17', 'Nuuk', Decimal('16.98')),
 (3, 'G', 'Rajkot', Decimal('16.98')),
 (13, 'G', 'Touliu', Decimal('16.98'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>San Bernardino</td>
        <td>24.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>Apeldoorn</td>
        <td>23.95</td>
    </tr>
    <tr>
        <td>21</td>
        <td>NC-17</td>
        <td>Belm</td>
        <td>22.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>R</td>
        <td>Mwanza</td>
        <td>21.97</td>
    </tr>
</tbody></table></div>

## Slicing

Slicing is the reduction of the dimensionality of a cube by 1 e.g. 3 dimensions to 2, fixing one of the dimensions to a single value. In the example above, we have a 3-dimensional cube on day, rating, and country.

Reduce the dimensionality of the above example by limiting the results to only include movies with a `rating` of "PG-13". Again, sort by revenue in descending order and limit to the first 20 rows.

In [3]:
%%time
%%sql

SELECT d.day AS day, m.rating AS rating, c.city AS city, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
WHERE m.rating='PG-13'
GROUP BY day, rating, city
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila
20 rows affected.
CPU times: user 6.92 ms, sys: 0 ns, total: 6.92 ms
Wall time: 38.2 ms


[(14, 'PG-13', 'Brockton', Decimal('21.97')),
 (15, 'PG-13', 'Jhansi', Decimal('20.97')),
 (10, 'PG-13', 'Uruapan', Decimal('17.97')),
 (1, 'PG-13', 's-Hertogenbosch', Decimal('16.98')),
 (8, 'PG-13', 'Probolinggo', Decimal('16.98')),
 (10, 'PG-13', 'Greensboro', Decimal('15.98')),
 (1, 'PG-13', 'Ogbomosho', Decimal('15.98')),
 (25, 'PG-13', 'Yantai', Decimal('15.98')),
 (22, 'PG-13', 'Jedda', Decimal('15.98')),
 (25, 'PG-13', 'Bhimavaram', Decimal('15.98')),
 (25, 'PG-13', 'Fukuyama', Decimal('15.97')),
 (6, 'PG-13', 'Ife', Decimal('14.98')),
 (21, 'PG-13', 'Otsu', Decimal('14.98')),
 (9, 'PG-13', 'Bhilwara', Decimal('14.98')),
 (14, 'PG-13', 'Fuyu', Decimal('14.98')),
 (25, 'PG-13', 'Omdurman', Decimal('14.98')),
 (4, 'PG-13', 'Molodetno', Decimal('14.98')),
 (22, 'PG-13', 'Bellevue', Decimal('14.98')),
 (3, 'PG-13', 'Ashdod', Decimal('14.98')),
 (27, 'PG-13', 'Binzhou', Decimal('14.98'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>PG-13</td>
        <td>Dhaka</td>
        <td>19.97</td>
    </tr>
    <tr>
        <td>29</td>
        <td>PG-13</td>
        <td>Shimoga</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Osmaniye</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>21</td>
        <td>PG-13</td>
        <td>Asuncin</td>
        <td>18.95</td>
    </tr>
</tbody></table></div>

## Dicing
Dicing is creating a subcube with the same dimensionality but fewer values for  two or more dimensions. 

Create a subcube of the initial cube that includes movies with:
* ratings of PG or PG-13
* in the city of Bellevue or Lancaster
* day equal to 1, 15, or 30

In [4]:
%%time
%%sql

SELECT d.day AS day, m.rating AS rating, c.city AS city, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
WHERE m.rating IN ('PG-13', 'PG') AND c.city IN ('Bellevue', 'Lancaster') AND d.day IN (1,15,30)
GROUP BY day, rating, city
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila
2 rows affected.
CPU times: user 7.48 ms, sys: 1.47 ms, total: 8.95 ms
Wall time: 16.7 ms


[(15, 'PG-13', 'Lancaster', Decimal('8.99')),
 (30, 'PG', 'Bellevue', Decimal('0.99'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>Lancaster</td>
        <td>12.98</td>
    </tr>
    <tr>
        <td>1</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>5.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>3.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>2.99</td>
    </tr>
    <tr>
        <td>15</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>1.98</td>
    </tr>
</tbody></table></div>

## 2. Roll-ups and Drill-downs

### Roll-up
- Stepping up the level of aggregation to a large grouping
- e.g.`city` is summed as `country`

Calculate revenue (sales_amount) by day, rating, and country. Sort the data by revenue in descending order, and limit the data to the top 20 results.

In [5]:
%%time
%%sql
SELECT d.day AS day, m.rating AS rating, c.country AS country, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
GROUP BY day, rating, country
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila
20 rows affected.
CPU times: user 6.06 ms, sys: 1.24 ms, total: 7.31 ms
Wall time: 63.1 ms


[(9, 'PG-13', 'India', Decimal('92.83')),
 (25, 'PG-13', 'China', Decimal('86.85')),
 (25, 'PG-13', 'India', Decimal('77.85')),
 (27, 'PG-13', 'China', Decimal('77.82')),
 (14, 'NC-17', 'India', Decimal('74.83')),
 (18, 'PG-13', 'India', Decimal('73.87')),
 (16, 'R', 'China', Decimal('73.85')),
 (12, 'NC-17', 'India', Decimal('73.82')),
 (27, 'R', 'India', Decimal('72.83')),
 (17, 'PG-13', 'India', Decimal('71.86')),
 (24, 'PG', 'United States', Decimal('71.85')),
 (26, 'PG', 'India', Decimal('71.83')),
 (14, 'PG-13', 'China', Decimal('70.85')),
 (14, 'R', 'India', Decimal('69.89')),
 (15, 'G', 'India', Decimal('69.84')),
 (24, 'R', 'China', Decimal('67.89')),
 (5, 'G', 'India', Decimal('67.87')),
 (5, 'PG-13', 'India', Decimal('67.84')),
 (4, 'PG', 'China', Decimal('66.86')),
 (17, 'PG-13', 'China', Decimal('66.85'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>China</td>
        <td>169.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>India</td>
        <td>156.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>India</td>
        <td>153.64</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>China</td>
        <td>146.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>R</td>
        <td>China</td>
        <td>145.66</td>
    </tr>
</tbody></table></div>

### Drill-down
- Breaking up one of the dimensions to a lower level.
- e.g.`city` is broken up into  `districts`

Calculate revenue (sales_amount) by day, rating, and district. Sort the data by revenue in descending order, and limit the data to the top 20 results.

In [6]:
%%time
%%sql
SELECT d.day AS day, m.rating AS rating, c.district AS district, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
GROUP BY day, rating, district
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila
20 rows affected.
CPU times: user 7.7 ms, sys: 194 µs, total: 7.9 ms
Wall time: 79.5 ms


[(18, 'PG-13', 'West Bengali', Decimal('36.95')),
 (27, 'NC-17', 'Buenos Aires', Decimal('36.91')),
 (19, 'G', 'California', Decimal('32.96')),
 (12, 'NC-17', 'Buenos Aires', Decimal('32.93')),
 (9, 'PG-13', 'Buenos Aires', Decimal('30.96')),
 (24, 'G', 'England', Decimal('30.95')),
 (24, 'PG-13', 'Southern Tagalog', Decimal('29.95')),
 (14, 'R', 'West Bengali', Decimal('28.95')),
 (8, 'G', 'Shandong', Decimal('27.97')),
 (18, 'G', 'Shandong', Decimal('27.96')),
 (3, 'R', 'West Bengali', Decimal('27.95')),
 (24, 'PG', 'California', Decimal('27.95')),
 (9, 'PG-13', 'Rajasthan', Decimal('26.96')),
 (3, 'PG-13', 'Southern Tagalog', Decimal('26.94')),
 (5, 'PG-13', 'Southern Tagalog', Decimal('25.97')),
 (12, 'G', 'Tamil Nadu', Decimal('25.96')),
 (16, 'NC-17', 'Shandong', Decimal('25.96')),
 (30, 'PG-13', 'West Bengali', Decimal('25.94')),
 (17, 'NC-17', 'Illinois', Decimal('24.97')),
 (25, 'NC-17', 'West Bengali', Decimal('24.97'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>district</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Southern Tagalog</td>
        <td>53.88</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Inner Mongolia</td>
        <td>38.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Shandong</td>
        <td>36.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>West Bengali</td>
        <td>36.92</td>
    </tr>
    <tr>
        <td>17</td>
        <td>PG-13</td>
        <td>Shandong</td>
        <td>34.95</td>
    </tr>
</tbody></table></div>

## 3. Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

### Revenue Total, by Month, by Country, by Month & Country All in one shot

The query calculates total revenue at the various grouping levels (total, by month, by country, by month & country) all at once using the grouping sets function.

In [7]:
%%sql
SELECT d.month AS month, s.country AS country, SUM(sales_amount) AS revenue
FROM factsales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimStore s ON f.store_key = s.store_key
GROUP BY GROUPING SETS ((), (month), (country), (month, country))

 * postgresql://student:***@127.0.0.1:5432/pagila
24 rows affected.


[(1, 'Australia', Decimal('1471.56')),
 (1, 'Canada', Decimal('1623.22')),
 (1, None, Decimal('3094.78')),
 (2, 'Australia', Decimal('5043.95')),
 (2, 'Canada', Decimal('5121.02')),
 (2, None, Decimal('10164.97')),
 (3, 'Australia', Decimal('5934.80')),
 (3, 'Canada', Decimal('5479.06')),
 (3, None, Decimal('11413.86')),
 (4, 'Australia', Decimal('5484.99')),
 (4, 'Canada', Decimal('5274.53')),
 (4, None, Decimal('10759.52')),
 (5, 'Australia', Decimal('5630.65')),
 (5, 'Canada', Decimal('5716.63')),
 (5, None, Decimal('11347.28')),
 (6, 'Australia', Decimal('5372.50')),
 (6, 'Canada', Decimal('5550.95')),
 (6, None, Decimal('10923.45')),
 (7, 'Australia', Decimal('4788.32')),
 (7, 'Canada', Decimal('4924.33')),
 (7, None, Decimal('9712.65')),
 (None, None, Decimal('67416.51')),
 (None, 'Canada', Decimal('33689.74')),
 (None, 'Australia', Decimal('33726.77'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>

## 4. CUBE Queries
- Group by CUBE (dim1, dim2, ..) , produces all combinations of different lengths in one go.
- This view could be materialized in a view and queried which would save lots repetitive aggregations

Calculate the various levels of aggregation done in the grouping sets exercise (total, by month, by country, by month & country) using the CUBE function.

In [8]:
%%time
%%sql
SELECT d.month AS month, s.country AS country, SUM(sales_amount) AS revenue
FROM factsales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimStore s ON f.store_key = s.store_key
GROUP BY CUBE (month, country)

 * postgresql://student:***@127.0.0.1:5432/pagila
24 rows affected.
CPU times: user 7.09 ms, sys: 1.43 ms, total: 8.52 ms
Wall time: 33.5 ms


[(1, 'Australia', Decimal('1471.56')),
 (1, 'Canada', Decimal('1623.22')),
 (1, None, Decimal('3094.78')),
 (2, 'Australia', Decimal('5043.95')),
 (2, 'Canada', Decimal('5121.02')),
 (2, None, Decimal('10164.97')),
 (3, 'Australia', Decimal('5934.80')),
 (3, 'Canada', Decimal('5479.06')),
 (3, None, Decimal('11413.86')),
 (4, 'Australia', Decimal('5484.99')),
 (4, 'Canada', Decimal('5274.53')),
 (4, None, Decimal('10759.52')),
 (5, 'Australia', Decimal('5630.65')),
 (5, 'Canada', Decimal('5716.63')),
 (5, None, Decimal('11347.28')),
 (6, 'Australia', Decimal('5372.50')),
 (6, 'Canada', Decimal('5550.95')),
 (6, None, Decimal('10923.45')),
 (7, 'Australia', Decimal('4788.32')),
 (7, 'Canada', Decimal('4924.33')),
 (7, None, Decimal('9712.65')),
 (None, None, Decimal('67416.51')),
 (None, 'Canada', Decimal('33689.74')),
 (None, 'Australia', Decimal('33726.77'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>